### PR TITLE
gNMI-1.2: Benchmarking: Full Configuration Replace 

### LINKED ISSUE #138: gNMI-1.2:  Benchmarking: Full Configuration Replace

### PR DESCRIPTION
**Pull request by yashikakn**
Resolves #138. Initial commit for "gNMI-1.2: Benchmarking: Full Configuration Replace ". Please review and provide approval. 
Thanks.

### CODE REVIEW

File: feature/experimental/system/gnmi/benchmarking/ate_tests/full_configuration_replace_test/full_configuration_replace_test.go [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+	gnmi.Replace(t, dut, dutConfNIPath.Type().Config(), oc.NetworkInstanceTypes_NETWORK_INSTANCE_TYPE_DEFAULT_INSTANCE)
+
+	t.Log("Cleanup exisitng bgp and isis configs on DUT before configuring test configs")
+	dutBGPPath := gnmi.OC().NetworkInstance(*deviations.DefaultNetworkInstance).Protocol(oc.PolicyTypes_INSTALL_PROTOCOL_TYPE_BGP, "BGP").Bgp()
+	gnmi.Delete(t, dut, dutBGPPath.Config())
+	dutISISPath := gnmi.OC().NetworkInstance(*deviations.DefaultNetworkInstance).Protocol(oc.PolicyTypes_INSTALL_PROTOCOL_TYPE_ISIS, setup.IsisInstance).Isis()
+	gnmi.Delete(t, dut, dutISISPath.Config())
+
+	t.Logf("Build interfaces, ISIS and BGP protocols configuration and send gNMI Set request")
+	setup.BuildOCUpdate(t)
```
Comment (xw-g) [inline feature/experimental/system/gnmi/benchmarking/ate_tests/full_configuration_replace_test/full_configuration_replace_test.go:94]: where is the time measurement of the full config push?

File: feature/experimental/system/gnmi/benchmarking/ate_tests/internal/setup/setup.go [hunk 2]
```diff
@@ -0,0 +1,324 @@
...
+	"github.com/openconfig/featureprofiles/internal/deviations"
+	"github.com/openconfig/featureprofiles/internal/fptest"
+	"github.com/openconfig/ondatra"
+	"github.com/openconfig/ondatra/gnmi"
+	"github.com/openconfig/ondatra/gnmi/oc"
+	"github.com/openconfig/ygnmi/ygnmi"
+	"github.com/openconfig/ygot/ygot"
+)
+
+// Some of the variables defined below like DutAS, AteAS, PeerGrpName
```
Comment (xw-g) [inline feature/experimental/system/gnmi/benchmarking/ate_tests/internal/setup/setup.go:38]: nit: no need to mention this. Golang naming can tell whether it's exported or not.

**Review by xw-g — COMMENTED** [review]
Besides a few code language related comments, I have a few thoughts overall: 

+ @liulk please help also comment my thoughts, as it's also related to infra.

* Different DUT target (or use case) might require different "maximum" number, therefore shall we consider making the `maximum` value from some input flags? (instead of statically coded?)
* Same argument would apply to the expected result (within how many seconds should the config push finish). 
* Seems like the new added internal/setup is specific to this test, and not generic enough to be used by other tests. If so, please make `setup.go` in the same package as this test.
* I think this test should goes to this folder `https://github.com/openconfig/featureprofiles/tree/main/feature/experimental/gnmi_service`

**Comment by liulk** [comment]
> Besides a few code language related comments, I have a few thoughts overall:
> 
> @liulk please help also comment my thoughts, as it's also related to infra.
> 
> * Different DUT target (or use case) might require different "maximum" number, therefore shall we consider making the `maximum` value from some input flags? (instead of statically coded?)
> * Same argument would apply to the expected result (within how many seconds should the config push finish).

I think the point of this test is to make sure full config replace does not take too long while exercising enough config items, so as long as the number is comfortably supported by all vendors, it can be hard coded. I would rather call it "num required" and avoid calling it "maximum" which can be confused with the device capability. Of course we still have to decide what that number should be, according to the program requirements. And how long is too long.

On the other hand, I could imagine a different type of test that checks that a device will indeed support X number of Y items (like you'd find in a datasheet). Such test will be parameterized by the content of the datasheet, and the device-specific parameters can be placed here: https://github.com/openconfig/featureprofiles/blob/main/internal/args/args.go

> * Seems like the new added internal/setup is specific to this test, and not generic enough to be used by other tests. If so, please make `setup.go` in the same package as this test.

I agree.

> * I think this test should goes to this folder `https://github.com/openconfig/featureprofiles/tree/main/feature/experimental/gnmi_service`

Why `gnmi_service` not just `gnmi`?

**Comment by liulk** [comment]
By the way, please rebase to the latest main and run `go run ./tools/addrundata --fix` after we decide on the final code location.

File: feature/experimental/system/gnmi/benchmarking/ate_tests/full_configuration_replace_test/README.md [hunk 3]
```diff
@@ -0,0 +1,27 @@
+# gNMI-1.2: Benchmarking: Full Configuration Replace 
+
+## Summary
+
+Measure performance of full configuration replace
```
Comment (greg-dennis) [inline feature/experimental/system/gnmi/benchmarking/ate_tests/full_configuration_replace_test/README.md:5]: Add period to end

File: feature/experimental/system/gnmi/benchmarking/ate_tests/full_configuration_replace_test/full_configuration_replace_test.go [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+
+	t.Logf("Modify description of a subset of interfaces and send gNMI Set request")
+	d2 := modIntfDesc(t)
+	confP := gnmi.OC()
+
+	fptest.LogQuery(t, "DUT", confP.Config(), d2)
+	//Start the timer.
```
Comment (greg-dennis) [inline feature/experimental/system/gnmi/benchmarking/ate_tests/full_configuration_replace_test/full_configuration_replace_test.go:101]: space after // here and below

File: feature/experimental/system/gnmi/benchmarking/ate_tests/full_configuration_replace_test/full_configuration_replace_test.go [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+	start := time.Now()
+
+	//conf.Update(t, d2)
```
Comment (greg-dennis) [inline feature/experimental/system/gnmi/benchmarking/ate_tests/full_configuration_replace_test/full_configuration_replace_test.go:104]: leftover debugging comment? delete

File: feature/experimental/system/gnmi/benchmarking/ate_tests/internal/setup/setup.go [hunk 6]
```diff
@@ -0,0 +1,324 @@
...
+// RouteCount and IsisInstance are used in other files which import
+// setup.go
+const (
+	// DutAS can be exported
```
Comment (greg-dennis) [inline feature/experimental/system/gnmi/benchmarking/ate_tests/internal/setup/setup.go:42]: The "can be exported" comments don't seem necessary, because they _are_ exported. Also, the constants IsisInstance and ISISMetric are exported but don't have the comment. Consider grouping the exported consts together.

File: feature/experimental/system/gnmi/benchmarking/ate_tests/internal/setup/setup.go [hunk 7]
```diff
@@ -0,0 +1,324 @@
...
+	PeerGrpName    = "BGP-PEER-GROUP"
+	plenIPv4       = 30
+	dutStartIPAddr = "192.0.2.1"
+	ateStartIPAddr = "192.0.2.2"
+	// RouteCount can be exported
+	RouteCount            = 200
+	advertiseBGPRoutesv4  = "203.0.113.1"
+	authPassword          = "ISISAuthPassword"
+	advertiseISISRoutesv4 = "198.18.0.0"
+	IsisInstance          = "DEFAULT"
```
Comment (greg-dennis) [inline feature/experimental/system/gnmi/benchmarking/ate_tests/internal/setup/setup.go:57]: If IsisInstance is meant to exported, rename to ISISInstance; otherwise lowercase

File: feature/experimental/system/gnmi/benchmarking/ate_tests/internal/setup/setup.go [hunk 8]
```diff
@@ -0,0 +1,324 @@
...
+	ISISMetric            = 100
+	// DUTAreaAddress can be exported
+	DUTAreaAddress = "49.0001"
+	// DUTSysID can be exported
+	DUTSysID = "1920.0000.2001"
+)
+
+var (
+	// DutIPPool can be exported
+	DutIPPool = make(map[string]net.IP)
```
Comment (greg-dennis) [inline feature/experimental/system/gnmi/benchmarking/ate_tests/internal/setup/setup.go:67]: The Go initialism rule says that these should be DUTIPPool and ATEIPPool

File: feature/experimental/system/gnmi/benchmarking/ate_tests/internal/setup/setup.go [hunk 9]
```diff
@@ -0,0 +1,324 @@
...
+
+// BuildOCUpdate function is to build  OC config for interfaces.
+// It reads ports from binding file and returns gpb update message
+// which will have configurations for all the ports.
+func BuildOCUpdate(t *testing.T) {
+	dut := ondatra.DUT(t, "dut")
+	d := &oc.Root{}
+
+	// Default Network instance and Global BGP configs
+	ni1 := d.GetOrCreateNetworkInstance(*deviations.DefaultNetworkInstance)
```
Comment (greg-dennis) [inline feature/experimental/system/gnmi/benchmarking/ate_tests/internal/setup/setup.go:120]: Hmm, "ni1" looks a lot like "nil" to my eye. Maybe just me ...

**Review by greg-dennis — CHANGES_REQUESTED** [review]

File: feature/experimental/system/gnmi/benchmarking/ate_tests/full_configuration_replace_test/full_configuration_replace_test.go [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+		if i%2 == 0 {
+			i := d.GetOrCreateInterface(dutPorts[i].Name())
+			i.Description = ygot.String("modified via oc")
+		}
+	}
+
+	return d
+}
+
+func TestGnmiFullConfigReplace(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (yashikakn) [reply feature/experimental/system/gnmi/benchmarking/ate_tests/full_configuration_replace_test/full_configuration_replace_test.go:79]: @xw-g : Thanks for reviewing.
1. Currently, the test picks up the number of interfaces to configure from the binding file. Same is used for configuring BGP peers and ISIS adjacencies. The test does not currently statically code the numbers. 
2. For expected result, I have added expected value to be passed as an input argument in  https://github.com/openconfig/featureprofiles/blob/main/internal/args/args.go per @liulk 's suggestion. 
3. setup.go is used by another test gNMI-1.3,  tracked through PR 645.
4. On the script location - will wait for confirmation before moving the script.

Also, added comment for func TestGnmiFullConfigReplace()

File: feature/experimental/system/gnmi/benchmarking/ate_tests/full_configuration_replace_test/full_configuration_replace_test.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (yashikakn) [reply feature/experimental/system/gnmi/benchmarking/ate_tests/full_configuration_replace_test/full_configuration_replace_test.go]: renamed BuildOCUpdate()  to BuildBenchmarkingConfig()

Comment (yashikakn) [reply feature/experimental/system/gnmi/benchmarking/ate_tests/full_configuration_replace_test/full_configuration_replace_test.go]: it was done in setup.go - moved the code to main test now.

File: feature/experimental/system/gnmi/benchmarking/ate_tests/internal/setup/setup.go (hunk 2 above)
Comment (yashikakn) [reply feature/experimental/system/gnmi/benchmarking/ate_tests/internal/setup/setup.go]: removed the comments

File: feature/experimental/system/gnmi/benchmarking/ate_tests/full_configuration_replace_test/README.md (hunk 3 above)
Comment (yashikakn) [reply feature/experimental/system/gnmi/benchmarking/ate_tests/full_configuration_replace_test/README.md]: done

File: feature/experimental/system/gnmi/benchmarking/ate_tests/full_configuration_replace_test/full_configuration_replace_test.go (hunk 4 above)
Comment (yashikakn) [reply feature/experimental/system/gnmi/benchmarking/ate_tests/full_configuration_replace_test/full_configuration_replace_test.go]: done

File: feature/experimental/system/gnmi/benchmarking/ate_tests/full_configuration_replace_test/full_configuration_replace_test.go (hunk 5 above)
Comment (yashikakn) [reply feature/experimental/system/gnmi/benchmarking/ate_tests/full_configuration_replace_test/full_configuration_replace_test.go]: taken care.

File: feature/experimental/system/gnmi/benchmarking/ate_tests/internal/setup/setup.go (hunk 6 above)
Comment (yashikakn) [reply feature/experimental/system/gnmi/benchmarking/ate_tests/internal/setup/setup.go]: removed comments

File: feature/experimental/system/gnmi/benchmarking/ate_tests/internal/setup/setup.go (hunk 8 above)
Comment (yashikakn) [reply feature/experimental/system/gnmi/benchmarking/ate_tests/internal/setup/setup.go]: renamed

File: feature/experimental/system/gnmi/benchmarking/ate_tests/internal/setup/setup.go (hunk 9 above)
Comment (yashikakn) [reply feature/experimental/system/gnmi/benchmarking/ate_tests/internal/setup/setup.go]: renamed

**Comment by yashikakn** [comment]
@xw-g @liulk @greg-dennis : Thank you all for reviewing. I have addressed all the review comments. Please take a look and let me know if anything.

**Comment by xw-g** [comment]
Because now we have @greg-dennis to help gate the code quality, I'll remove myself from the reviewers.

File: feature/experimental/system/gnmi/benchmarking/ate_tests/full_configuration_replace_test/README.md [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+Configure DUT with:
+ - Maximum number of interfaces to be supported.
+ - Maximum number of BGP peers to be supported.
+ - Maximum number of IS-IS adjacencies to be supported.
+Measure time required for Set operation to complete. 
+Modify descriptions of a subset of interfaces within the system.
+Measure time for Set to complete.
+
+Notes:
+This test does not cover entirely converged system, simply replacing
```
Comment (greg-dennis) [inline feature/experimental/system/gnmi/benchmarking/ate_tests/full_configuration_replace_test/README.md:18]: This sentence is a little hard to parse. Maybe something like "This test does not measure an the time to an entirely converged state, only to completion of the gNMI update." ?

**Review by greg-dennis — CHANGES_REQUESTED** [review]

File: feature/experimental/system/gnmi/benchmarking/ate_tests/internal/setup/setup.go [hunk 12]
```diff
@@ -0,0 +1,295 @@
...
+		isisIntfLevelAfi.Enabled = ygot.Bool(true)
+	}
+	p := gnmi.OC()
+	fptest.LogQuery(t, "DUT", p.Config(), d)
+
+	return d
+}
+
+// ConfigureATE configures ATE ports with IPv4, ISIS and BGP peers.
+func ConfigureATE(t *testing.T, ate *ondatra.ATEDevice) {
```
[the remark this answers is not in the document]
Comment (yashikakn) [reply feature/experimental/system/gnmi/benchmarking/ate_tests/internal/setup/setup.go:204]: These are used in another test gNMI-1.3 (PR #645)

File: feature/experimental/system/gnmi/benchmarking/ate_tests/internal/setup/setup.go [hunk 13]
```diff
@@ -0,0 +1,295 @@
...
+	"github.com/openconfig/ondatra"
+	"github.com/openconfig/ondatra/gnmi"
+	"github.com/openconfig/ondatra/gnmi/oc"
+	"github.com/openconfig/ygnmi/ygnmi"
+	"github.com/openconfig/ygot/ygot"
+)
+
+const (
+	ISISInstance          = "DEFAULT"
+	ISISMetric            = 100
```
[the remark this answers is not in the document]
Comment (yashikakn) [reply feature/experimental/system/gnmi/benchmarking/ate_tests/internal/setup/setup.go:40]: These are used in another test gNMI-1.3 (PR #645)

File: feature/experimental/system/gnmi/benchmarking/ate_tests/full_configuration_replace_test/full_configuration_replace_test.go [hunk 14]
```diff
@@ -0,0 +1,132 @@
...
+	"github.com/openconfig/ondatra/gnmi"
+	"github.com/openconfig/ondatra/gnmi/oc"
+	"github.com/openconfig/ygot/ygot"
+)
+
+func TestMain(m *testing.M) {
+	fptest.RunTests(m)
+}
+
+// Test case:
```
[the remark this answers is not in the document]
Comment (yashikakn) [reply feature/experimental/system/gnmi/benchmarking/ate_tests/full_configuration_replace_test/full_configuration_replace_test.go]: Agree. Removed them.

File: feature/experimental/system/gnmi/benchmarking/ate_tests/full_configuration_replace_test/README.md (hunk 11 above)
Comment (yashikakn) [reply feature/experimental/system/gnmi/benchmarking/ate_tests/full_configuration_replace_test/README.md]: Re-worded it per suggestion. Thanks.

File: feature/experimental/system/gnmi/benchmarking/ate_tests/internal/setup/setup.go [hunk 15]
```diff
@@ -0,0 +1,295 @@
...
+	"github.com/openconfig/featureprofiles/internal/fptest"
+	"github.com/openconfig/ondatra"
+	"github.com/openconfig/ondatra/gnmi"
+	"github.com/openconfig/ondatra/gnmi/oc"
+	"github.com/openconfig/ygnmi/ygnmi"
+	"github.com/openconfig/ygot/ygot"
+)
+
+const (
+	ISISInstance          = "DEFAULT"
```
[the remark this answers is not in the document]
Comment (yashikakn) [reply feature/experimental/system/gnmi/benchmarking/ate_tests/internal/setup/setup.go]: added comments.

File: feature/experimental/system/gnmi/benchmarking/ate_tests/full_configuration_replace_test/full_configuration_replace_test.go [hunk 16]
```diff
@@ -0,0 +1,117 @@
...
+// sortPorts sorts the ports by the testbed port ID.
+func sortPorts(ports []*ondatra.Port) []*ondatra.Port {
+	sort.SliceStable(ports, func(i, j int) bool {
+		return ports[i].ID() < ports[j].ID()
+	})
+	return ports
+}
+
+// modifyIntfDescription builds config to modify description of a subset of interfaces.
+func modifyIntfDescription(t *testing.T) *oc.Root {
```
Comment (greg-dennis) [inline feature/experimental/system/gnmi/benchmarking/ate_tests/full_configuration_replace_test/full_configuration_replace_test.go:48]: Change this method to accept (dut *ondatra.DUTDevice) instead of (t *testing.T). No need to repeat the ondatra.DUT lookup the caller is already doing and no need to have the implicit dependency on the name "dut" here

File: feature/experimental/system/gnmi/benchmarking/ate_tests/full_configuration_replace_test/full_configuration_replace_test.go [hunk 17]
```diff
@@ -0,0 +1,117 @@
...
+	"github.com/openconfig/ondatra/gnmi/oc"
+	"github.com/openconfig/ygot/ygot"
+)
+
+func TestMain(m *testing.M) {
+	fptest.RunTests(m)
+}
+
+// Topology:
+// 	dut:port(1..N)
```
Comment (greg-dennis) [inline feature/experimental/system/gnmi/benchmarking/ate_tests/full_configuration_replace_test/full_configuration_replace_test.go:37]: This two-line comment is a little unclear about what it is communicating. That there is some N number of ports? I suspect it's duplicative with whatever is in the README (or should be in the README) and can be safely dropped here.

File: feature/experimental/system/gnmi/benchmarking/ate_tests/internal/setup/setup.go [hunk 18]
```diff
@@ -0,0 +1,296 @@
...
+	"github.com/openconfig/featureprofiles/internal/fptest"
+	"github.com/openconfig/ondatra"
+	"github.com/openconfig/ondatra/gnmi"
+	"github.com/openconfig/ondatra/gnmi/oc"
+	"github.com/openconfig/ygnmi/ygnmi"
+	"github.com/openconfig/ygot/ygot"
+)
+
+const (
+	ISISInstance   = "DEFAULT"        // ISIS instance name.
```
Comment (greg-dennis) [inline feature/experimental/system/gnmi/benchmarking/ate_tests/internal/setup/setup.go:39]: Please fix all the static analysis failures on this file. I'd move the comment to the line immediately above the comment initialization and follow the style guide format as described in the static analysis error, e.g.

// ISISInstance is [fill in the description here].

File: feature/experimental/system/gnmi/benchmarking/ate_tests/internal/setup/setup.go [hunk 19]
```diff
@@ -0,0 +1,296 @@
...
+var (
+	DUTIPList       = make(map[string]net.IP) // DUT IP list.
+	ATEIPList       = make(map[string]net.IP) // ATE IP list.
+	ISISMetricArray []uint32                  // ISIS metric array.
+	ISISSetBitArray []bool                    // ISIS set bit array.
+)
+
+// BuildIPList builds list of ip addresses for the ports in binding file.
+// (Both DUT and ATE ports).
+func BuildIPList(t *testing.T) {
```
Comment (greg-dennis) [inline feature/experimental/system/gnmi/benchmarking/ate_tests/internal/setup/setup.go:66]: This method can accept (dut *ondatra.DUTDevice) instead of testing.T

File: feature/experimental/system/gnmi/benchmarking/ate_tests/internal/setup/setup.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (greg-dennis) [reply feature/experimental/system/gnmi/benchmarking/ate_tests/internal/setup/setup.go]: Ideally the helpers are added in the PR in which they are used, because that's the time when they can be reviewed to make sure they are fit for purpose. Can that be done in this case -- add the helpers that are used by that PR in that PR?

File: feature/experimental/system/gnmi/benchmarking/ate_tests/internal/setup/setup.go [hunk 20]
```diff
@@ -0,0 +1,296 @@
...
+
+	return d
+}
+
+// ConfigureATE configures ATE ports with IPv4, ISIS and BGP peers.
+func ConfigureATE(t *testing.T, ate *ondatra.ATEDevice) {
+	topo := ate.Topology().New()
+
+	for _, dp := range ate.Ports() {
+		ISISMetricArray = append(ISISMetricArray, ISISMetric)
```
Comment (greg-dennis) [inline feature/experimental/system/gnmi/benchmarking/ate_tests/internal/setup/setup.go:209]: I don't see the need to export ISISMetricArray or ISISSetBitArray. They repeat the same values for the number of ports, but that may just be simpler to let the caller do. (Again, hard to evaluate because there are no users of this field in the PR).

**Review by greg-dennis — CHANGES_REQUESTED** [review]

**Comment by liulk** [comment]
> > * I think this test should goes to this folder `https://github.com/openconfig/featureprofiles/tree/main/feature/experimental/gnmi_service`
> 
> Why `gnmi_service` not just `gnmi`?

I talked to @sachendras and we agreed that the location is fine for now. Okay for this PR to use `experimental/system/gnmi`.

I think @xw-g is concerned about reusing the existing `experimental/gnmi_service` location and avoid creating a new `experimental/system/gnmi` location, I'm mostly concerned about why not use `experimental/gnmi` but it will require some cleanup.

Once this PR is merged, I'll consolidate the test locations. Thanks!

File: feature/experimental/system/gnmi/benchmarking/ate_tests/internal/setup/setup.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (yashikakn) [reply feature/experimental/system/gnmi/benchmarking/ate_tests/internal/setup/setup.go]: Removed all the content which is not needed by this PR. Please take a look.

File: feature/experimental/system/gnmi/benchmarking/ate_tests/full_configuration_replace_test/full_configuration_replace_test.go (hunk 16 above)
Comment (yashikakn) [reply feature/experimental/system/gnmi/benchmarking/ate_tests/full_configuration_replace_test/full_configuration_replace_test.go]: Done. Thanks.

File: feature/experimental/system/gnmi/benchmarking/ate_tests/internal/setup/setup.go (hunk 18 above)
Comment (yashikakn) [reply feature/experimental/system/gnmi/benchmarking/ate_tests/internal/setup/setup.go]: Done.

File: feature/experimental/system/gnmi/benchmarking/ate_tests/internal/setup/setup.go (hunk 19 above)
Comment (yashikakn) [reply feature/experimental/system/gnmi/benchmarking/ate_tests/internal/setup/setup.go]: Done.

File: feature/experimental/system/gnmi/benchmarking/ate_tests/internal/setup/setup.go (hunk 20 above)
Comment (yashikakn) [reply feature/experimental/system/gnmi/benchmarking/ate_tests/internal/setup/setup.go]: Resolved by removing all the content which is not needed by this PR.

File: feature/experimental/system/gnmi/benchmarking/ate_tests/full_configuration_replace_test/full_configuration_replace_test.go (hunk 17 above)
Comment (yashikakn) [reply feature/experimental/system/gnmi/benchmarking/ate_tests/full_configuration_replace_test/full_configuration_replace_test.go:37]: I have added a note for user convenience. Please see if it's clear now.

**Review by sachendras — APPROVED** [review]

File: feature/experimental/system/gnmi/benchmarking/ate_tests/full_configuration_replace_test/README.md [hunk 21]
```diff
@@ -0,0 +1,25 @@
...
+## Summary
+
+Measure performance of full configuration replace.
+
+## Procedure
+
+Configure DUT with:
+ - Maximum number of interfaces to be supported.
+ - Maximum number of BGP peers to be supported.
+ - Maximum number of IS-IS adjacencies to be supported.
```
Comment (LimeHat) [inline feature/experimental/system/gnmi/benchmarking/ate_tests/full_configuration_replace_test/README.md:12]: I find these statements about "maximum number of ..." slightly confusing. In the code you configure 1 BGP neighbor & 1 ISIS adjacency per ATE interface. Can we rephrase it (to avoid potential confusion with scalability testing / maximum number of supported interfaces/adjacencies on a node)?

File: feature/experimental/system/gnmi/benchmarking/ate_tests/internal/setup/setup.go [hunk 22]
```diff
@@ -0,0 +1,200 @@
...
+	isisTimers.LspLifetimeInterval = ygot.Uint16(600)
+	spfTimers := isisTimers.GetOrCreateSpf()
+	spfTimers.SpfHoldInterval = ygot.Uint64(5000)
+	spfTimers.SpfFirstInterval = ygot.Uint64(600)
+
+	isisLevel1 := isis.GetOrCreateLevel(1)
+	isisLevel1.Enabled = ygot.Bool(false)
+
+	isisLevel2 := isis.GetOrCreateLevel(2)
+	isisLevel2.Enabled = ygot.Bool(true)
```
Comment (LimeHat) [inline feature/experimental/system/gnmi/benchmarking/ate_tests/internal/setup/setup.go:136]: This leaf is soon to be deprecated in OpenConfig model, can we replace it with level-capability configuration?
ref. openconfig/public/pull/794

Comment (yashikakn) [reply feature/experimental/system/gnmi/benchmarking/ate_tests/internal/setup/setup.go]: @LimeHat : Thanks for the review. I have made the changes. Please take a look.

File: feature/experimental/system/gnmi/benchmarking/ate_tests/full_configuration_replace_test/README.md (hunk 21 above)
Comment (yashikakn) [reply feature/experimental/system/gnmi/benchmarking/ate_tests/full_configuration_replace_test/README.md]: Updated the description; please see if it is ok.

File: feature/experimental/system/gnmi/benchmarking/ate_tests/full_configuration_replace_test/full_configuration_replace_test.go (hunk 17 above)
Comment (greg-dennis) [reply feature/experimental/system/gnmi/benchmarking/ate_tests/full_configuration_replace_test/full_configuration_replace_test.go:37]: I don't think the comments really add much beyond what is already understood about the nature of these tests. I'd just drop.

File: feature/experimental/system/gnmi/benchmarking/ate_tests/internal/setup/setup.go [hunk 23]
```diff
@@ -0,0 +1,197 @@
...
+	"github.com/openconfig/ondatra/gnmi"
+	"github.com/openconfig/ondatra/gnmi/oc"
+	"github.com/openconfig/ygot/ygot"
+)
+
+const (
+	// ISISInstance is ISIS instance name.
+	ISISInstance = "DEFAULT"
+	// DUTAreaAddress is DUT ISIS area address.
+	DUTAreaAddress = "49.0001"
```
Comment (greg-dennis) [inline feature/experimental/system/gnmi/benchmarking/ate_tests/internal/setup/setup.go:39]: Looks like the only constant that is used outside the package is ISISInstance. The others are only used by this package and so don't need to be exported

File: feature/experimental/system/gnmi/benchmarking/ate_tests/internal/setup/setup.go [hunk 24]
```diff
@@ -0,0 +1,197 @@
...
+var (
+	// DUTIPList is DUT IP list.
+	DUTIPList = make(map[string]net.IP)
+	// ATEIPList is ATE IP list.
+	ATEIPList = make(map[string]net.IP)
+)
+
+// BuildIPList builds list of ip addresses for the ports in binding file.
+// (Both DUT and ATE ports).
+func BuildIPList(dut *ondatra.DUTDevice) {
```
Comment (greg-dennis) [inline feature/experimental/system/gnmi/benchmarking/ate_tests/internal/setup/setup.go:65]: To things:
- Rather than modify global variables, it's safer to have this function return the IPLists.
- Those IP lists look to only be used by BuildBenchmarkingConfig, so I think makes sense to un-export this and have it called by BuildBenchmarkingConfig

**Review by greg-dennis — CHANGES_REQUESTED** [review]

File: feature/experimental/system/gnmi/benchmarking/ate_tests/internal/setup/setup.go (hunk 23 above)
Comment (yashikakn) [reply feature/experimental/system/gnmi/benchmarking/ate_tests/internal/setup/setup.go]: Removed the comment. And double checked on exported const and updated the file.

File: feature/experimental/system/gnmi/benchmarking/ate_tests/internal/setup/setup.go (hunk 24 above)
Comment (yashikakn) [reply feature/experimental/system/gnmi/benchmarking/ate_tests/internal/setup/setup.go]: Changes done. Please take a look. Thanks.

File: feature/experimental/system/gnmi/benchmarking/ate_tests/full_configuration_replace_test/full_configuration_replace_test.go (hunk 17 above)
Comment (yashikakn) [reply feature/experimental/system/gnmi/benchmarking/ate_tests/full_configuration_replace_test/full_configuration_replace_test.go]: Removed the comment.

File: feature/experimental/system/gnmi/benchmarking/ate_tests/internal/setup/setup.go [hunk 25]
```diff
@@ -0,0 +1,194 @@
...
+	"github.com/openconfig/ondatra/gnmi"
+	"github.com/openconfig/ondatra/gnmi/oc"
+	"github.com/openconfig/ygot/ygot"
+)
+
+const (
+	// ISISInstance is ISIS instance name.
+	ISISInstance = "DEFAULT"
+	// PeerGrpName is BGP peer group name.
+	PeerGrpName = "BGP-PEER-GROUP"
```
Comment (greg-dennis) [inline feature/experimental/system/gnmi/benchmarking/ate_tests/internal/setup/setup.go:39]: not used outside the package, un-export it

File: feature/experimental/system/gnmi/benchmarking/ate_tests/internal/setup/setup.go [hunk 26]
```diff
@@ -0,0 +1,194 @@
...
+	// DUTAs is DUT AS.
+	DUTAs = 64500
```
Comment (greg-dennis) [inline feature/experimental/system/gnmi/benchmarking/ate_tests/internal/setup/setup.go:41]: not used outside the package, un-export it

File: feature/experimental/system/gnmi/benchmarking/ate_tests/internal/setup/setup.go [hunk 27]
```diff
@@ -0,0 +1,194 @@
...
+	// ATEAs is ATE AS.
+	ATEAs = 64501
```
Comment (greg-dennis) [inline feature/experimental/system/gnmi/benchmarking/ate_tests/internal/setup/setup.go:43]: not used outside the package, un-export it

File: feature/experimental/system/gnmi/benchmarking/ate_tests/internal/setup/setup.go [hunk 28]
```diff
@@ -0,0 +1,194 @@
...
+	dutAreaAddress = "49.0001"
+	dutSysID       = "1920.0000.2001"
+	ateAs2         = 64502
+	dutStartIPAddr = "192.0.2.1"
+	ateStartIPAddr = "192.0.2.2"
+	plenIPv4       = 30
+	authPassword   = "ISISAuthPassword"
+)
+
+// buildIPList builds list of ip addresses for the ports in binding file.
```
Comment (greg-dennis) [inline feature/experimental/system/gnmi/benchmarking/ate_tests/internal/setup/setup.go:54]: naming and documentation is a little unclear, because it says it builds a "list" but it actually builds two maps. In addition to correcting the godoc, maybe this should be named buildPortIPLists (plural)? And then consider pluralizing the variables dutIPList and ateIPList to dutIPLists and ateIPLists

**Review by greg-dennis — CHANGES_REQUESTED** [review]

File: feature/experimental/system/gnmi/benchmarking/ate_tests/internal/setup/setup.go (hunk 25 above)
Comment (yashikakn) [reply feature/experimental/system/gnmi/benchmarking/ate_tests/internal/setup/setup.go:39]: This is exported by PR #645, so just left it as it is.

File: feature/experimental/system/gnmi/benchmarking/ate_tests/internal/setup/setup.go (hunk 26 above)
Comment (yashikakn) [reply feature/experimental/system/gnmi/benchmarking/ate_tests/internal/setup/setup.go:41]: This is exported by PR #645, so just left it as it is.

File: feature/experimental/system/gnmi/benchmarking/ate_tests/internal/setup/setup.go (hunk 27 above)
Comment (yashikakn) [reply feature/experimental/system/gnmi/benchmarking/ate_tests/internal/setup/setup.go:43]: This is used by PR #645 also, so just left it as it is.

File: feature/experimental/system/gnmi/benchmarking/ate_tests/internal/setup/setup.go (hunk 25 above)
Comment (greg-dennis) [reply feature/experimental/system/gnmi/benchmarking/ate_tests/internal/setup/setup.go:39]: Will let these through as is. In the future, better to leave it unexported in the PR that introduces it and change it to exported in the PR that needs it to be.

Comment (yashikakn) [reply feature/experimental/system/gnmi/benchmarking/ate_tests/internal/setup/setup.go:39]: Point noted. Thanks Greg.

File: feature/experimental/system/gnmi/benchmarking/ate_tests/internal/setup/setup.go (hunk 28 above)
Comment (yashikakn) [reply feature/experimental/system/gnmi/benchmarking/ate_tests/internal/setup/setup.go]: Re-worded comment and re-named variables. Please see if it's ok. Thanks.

**Review by greg-dennis — APPROVED** [review]

**Comment by sachendras** [comment]
Google's internal thread for reference b/265859400

**Review by sachendras — APPROVED** [review]
